### ISSUE TITLE
Uncaught RangeError: Maximum call stack size exceeded

### ISSUE THREAD

**Issue by gytdau** [PrismJS/prism#818]
This error occurs in my console when the script is included, even if I didn't add any language classes anywhere.

This is my body:
```
<body>
<pre><code>
Code goes here
</code></pre>
<script src="/prism/prism.js"></script>
</body>
```

I've included most of the languages available in my installation, and the line numbers plugin.

This is the full error, as seen on my Chrome dev tools:
```
Uncaught RangeError: Maximum call stack size exceeded
match @ (program):128
_self.Prism.util.type @ prism.js:2
_self.Prism.languages.DFS @ prism.js:2
_self.Prism.languages.DFS @ prism.js:2
_self.Prism.languages.DFS @ prism.js:2
_self.Prism.languages.DFS @ prism.js:2
_self.Prism.languages.DFS @ prism.js:2
_self.Prism.languages.DFS @ prism.js:2
_self.Prism.languages.DFS @ prism.js:2
_self.Prism.languages.DFS @ prism.js:2
```

The highlighting still works - I just can't get the line numbers to work when I add a 'line-number' class to the `<pre>`.

**Comment by Golmote** [comment]
Hi, thanks for reporting.

This issue will be a tough one to fix... The DFS function is not able anymore to walk through all the languages. And the DFS function is called anytime insertBefore is called, which means... many times.

The DFS function is mainly needed because Prism relies on the order of the keys in the grammars. insertBefore creates a new object each time it's called (it can't be done "in place") and it needs to walk through all references of the previous object to update each one to the new reference.

I think #799 could help fixing this. I need to get back working on it.

For now, the only solution I can suggest is that you reduce the number of languages you included in your installation, preferably to match only the ones you actually need.

**Comment by gytdau** [comment]
After some testing, I can indeed confirm that the problem stems from too many languages. 

Thanks for your support!

**Comment by bebraw** [comment]
FYI, I hit the same problem at [prism-languages](https://github.com/survivejs/prism-languages) when trying to get it up to date. Given I just concat the language definitions, this happens.

**Comment by bebraw** [comment]
I "resolved" this problem at **prism-languages**. I simply hacked together a dependency declaration of my own. Now it's possible to consume the definitions from Node easily. This makes it ideal for backend processing. You probably don't want to use this on frontend, though, due to the sheer amount of definitions.

Ideally we would be able to drop *prism-languages* altogether. The current Node version of Prism seems better than what it used to be, but it's still missing most fo the language definitions.

**Comment by digitalmoksha** [comment]
I don't think the cause is from too many languages.  I started running into this same problem when I wanted to add a couple languages, and of course pulled the very latest js code.

There is something about a specific language module that seems to cause the problem.  `haml` is one of them.  For example, this combination works:

http://prismjs.com/download.html?themes=prism&languages=markup+css+clike+javascript+actionscript+apacheconf+applescript+aspnet+bash+c+csharp+cpp+coffeescript+ruby+css-extras+dart+git+go+handlebars+java+latex+less+lua+makefile+markdown+nginx+objectivec+perl+php+php-extras+python+r+scala+sql+swift+wiki+yaml

Add `haml` to it, and you get the error.  It was through the process of elimination that I found `haml` was the problem.  So even the following minimal set fails:

http://prismjs.com/download.html?themes=prism&languages=markup+css+clike+javascript+ruby+haml

There is at least one other language (not in my normal use set) that is causing the same problem, because if I include all languages except haml, the error occurs.

So the solution for me at the moment is not to include haml. I don't have the time to try and figure out exactly why haml causes the failure, but maybe this points someone in the right direction.

**Comment by iboshkov** [comment]
For some reason HAML didn't trigger the error for me, it was AsciiDoc that did it. Not sure why, at all.

I included all languages a few at a time (including haml), and the only one that caused the error was AsciiDoc.

**Comment by zeitgeist87** [comment]
It is not the number of languages that causes the problem. AsciiDoc contains an infinite recursion and DFS runs into it if you include any language with a call to insertBefore() after AsciiDoc.

AsciiDoc includes the following pattern:
```js
Prism.languages.asciidoc = {
'macro': {
			pattern: /\b[a-z\d][a-z\d-]*::?(?:(?:\S+)??\[(?:[^\]\\"]|(["'])(?:(?!\1)[^\\]|\\.)*\1|\\.)*\])/,
			inside: {
				'function': /^[a-z\d-]+(?=:)/,
				'punctuation': /^::?/,
				'attributes': {
					pattern: /(?:\[(?:[^\]\\"]|(["'])(?:(?!\1)[^\\]|\\.)*\1|\\.)*\])/,
					inside: attributes.inside
				}
			}
		},
```

**Comment by zeitgeist87** [comment]
The simplest way to reproduce the problem is to select AsciiDoc first and then for example Ruby:

http://prismjs.com/download.html?themes=prism&languages=asciidoc+ruby

**Comment by zeitgeist87** [comment]
I have quickly put together a PR to prevent infinite recursion in DFS #841. That PR would also solve this issue, but it is probably better to fix AsciiDoc instead.

**Comment by zeitgeist87** [comment]
The following patch solves this issue, but it kind of breaks the AsciiDoc language. The cloned versions of `macro` and `inline` don't have all the fields, because that would cause an infinite recursion.

```patch
diff --git a/components/prism-asciidoc.js b/components/prism-asciidoc.js
index 55a6330..63bb51b 100644
--- a/components/prism-asciidoc.js
+++ b/components/prism-asciidoc.js
@@ -200,8 +200,8 @@
        // Allow some nesting. There is no recursion though, so cloning should not be needed.
 
        attributes.inside['interpreted'].inside.rest = {
-               'macro': Prism.languages.asciidoc['macro'],
-               'inline': Prism.languages.asciidoc['inline'],
+               'macro': Prism.util.clone(Prism.languages.asciidoc['macro']),
+               'inline': Prism.util.clone(Prism.languages.asciidoc['inline']),
                'replacement': Prism.languages.asciidoc['replacement'],
                'entity': Prism.languages.asciidoc['entity']
        };
@@ -238,7 +238,7 @@
        };
 
        Prism.languages.asciidoc['other-block'].inside.rest = {
-               'table': Prism.languages.asciidoc['table'],
+               'table': Prism.util.clone(Prism.languages.asciidoc['table']),
                'list-punctuation': Prism.languages.asciidoc['list-punctuation'],
                'indented-block': Prism.languages.asciidoc['indented-block'],
                'comment': Prism.languages.asciidoc['comment'],
```

I think the better solution is #841, because it enables such complex self-referencing languages like AsciiDoc.

**Comment by zeitgeist87** [comment]
This issue was fixed with #841. Closing.
